contracts/EuroToken.sol

```diff
@@ -0,0 +1,235 @@
+contract EuroToken is
+    ITokenWithDeposit,
+    IERC677Token,
+    AccessControlled,
+    StandardToken,
+    TokenMetadata,
+    MigrationSource,
+    Reclaimable
+{
+    ////////////////////////
+    // Constants
+    ////////////////////////
+
+    string private constant NAME = "Euro Token";
+
+    string private constant SYMBOL = "EUR-T";
+
+    uint8 private constant DECIMALS = 18;
+
+    ////////////////////////
+    // Mutable state
+    ////////////////////////
+
+    // a list of addresses that are allowed to receive EUR-T
+    mapping(address => bool) private _allowedTransferTo;
+
+    // a list of of addresses that are allowed to send EUR-T
+    mapping(address => bool) private _allowedTransferFrom;
+
+    ////////////////////////
+    // Events
+    ////////////////////////
+
+    event AllowedFromAddress(
+        address indexed from,
+        bool allowed
+    );
+
+    event AllowedToAddress(
+        address indexed to,
+        bool allowed
+    );
+
+    /// @notice logged on successful migration
+    event OwnerMigrated(
+        address indexed owner,
+        uint256 amount
+    );
+
+    ////////////////////////
+    // Modifiers
+    ////////////////////////
+
+    modifier onlyAllowedTransferFrom(address from) {
+        require(_allowedTransferFrom[from]);
+        _;
+    }
+
+    modifier onlyAllowedTransferTo(address to) {
+        require(_allowedTransferTo[to]);
+        _;
+    }
+
+    ////////////////////////
+    // Constructor
+    ////////////////////////
+
+    function EuroToken(IAccessPolicy accessPolicy)
+        AccessControlled(accessPolicy)
+        StandardToken()
+        TokenMetadata(NAME, DECIMALS, SYMBOL, "")
+        MigrationSource(accessPolicy, ROLE_EURT_DEPOSIT_MANAGER)
+        Reclaimable()
+    {
+    }
+
+    ////////////////////////
+    // Public functions
+    ////////////////////////
+
+    /// @notice deposit 'amount' of EUR-T to address 'to'
+    /// @dev address 'to' is whitelisted as recipient of future transfers
+    /// @dev deposit may happen only in case of succesful KYC of recipient and validation of banking data
+    /// @dev which in this implementation is an off-chain responsibility of EURT_DEPOSIT_MANAGER
+    function deposit(address to, uint256 amount)
+        public
+        payable
+        only(ROLE_EURT_DEPOSIT_MANAGER)
+        returns (bool)
+    {
+        // cannot receive ether
+        require(msg.value == 0);
+        _balances[to] = add(_balances[to], amount);
+        _totalSupply = add(_totalSupply, amount);
+        setAllowedTransferTo(to, true);
+        Deposit(to, amount);
+        return true;
+    }
+
+    /// @notice withdraws 'amount' of EUR-T by burning required amount and providing a proof of whithdrawal
+    /// @dev proof is provided in form of log entry on which EURT_DEPOSIT_MANAGER
+    /// @dev will act off-chain to return required Euro amount to EUR-T holder
+    function withdraw(uint256 amount)
+        public
```

Comment:
We may want to put a role on here too, so we can enable/disable withdraws if we need to.

Reply:
OK